### PR TITLE
Model selection

### PR DESCRIPTION
**Pull request by VinsRR**
This branch proposes the following modifications:
- Implementation of model selection for MultiOrderModel (for walk data only)
- tests for the modules related to model selection in MultiOrderModel
- functions `compute_weighted_outdegree` and `compute_transition_probabilities` (necessary to perform model selection. These functions are located in MultiOrderModel.py
- fixing weights for higher-order networks (thus making the branch `fix-hon-weights` obsolete) #157 

The modifications in this branch also relate to the following topics:
- Edge weight representation #155 -- workarounds have been developed to relate edge frequencies to path frequencies (as, currently, path frequencies are saved into edge frequencies). See all the `frequencies` variables in MultiOrderModel  
- Zeroth order layer:  workarounds have been developed to perform the model selection without needing the zeroth order in MultiOrderModel. The question remains whether we want to implement one or not #172

### CODE REVIEW

File: src/pathpyG/core/MultiOrderModel.py [hunk 1]
```diff
@@ -278,3 +282,280 @@ def from_DAGs(
...
+        # Get ixs starting nodes
+        # Q: Is dag_graph.ptr[:-1] enough to get the start_ixs?
+        mask = torch.ones(dag_graph.num_nodes, dtype=bool)
+        mask[dag_graph.edge_index[1]] = False
+        start_ixs = dag_graph.node_sequence.squeeze()[mask]
+
+        # Compute node emission probabilities
+        # TODOL modify once we have zeroth order in mon
+        _, counts = torch.unique(dag_graph.node_sequence, return_counts=True)
+        node_emission_probabilities = counts / counts.sum()
```
[the remark this answers is not in the document]
Comment (M-Lampert) [reply src/pathpyG/core/MultiOrderModel.py:380]: We discussed this in our meeting and agreed to ignore it intentionally. I still think we should at least add it as comment so that we won't forget about it.

File: src/pathpyG/core/MultiOrderModel.py [hunk 2]
```diff
@@ -278,3 +282,280 @@ def from_DAGs(
...
+                    compute_transition_probabilities(self.layers[order])[
+                        self.layers[order + 1].data.inverse_idx[ho_intermediate_ixs]
+                    ]
+                ),
+            )
+            .sum()
+            .item()
+        )
+
+    def get_mon_log_likelihood(self, dag_graph: DataBatch, max_order: int = 1) -> float:
```
Comment (M-Lampert) [inline src/pathpyG/core/MultiOrderModel.py:421]:
```suggestion
    def get_mon_log_likelihood(self, dag_graph: Data, max_order: int = 1) -> float:
```

File: src/pathpyG/core/MultiOrderModel.py [hunk 3]
```diff
@@ -278,3 +282,280 @@ def from_DAGs(
...
+        # Adding the likelihood of highest/stationary order
+        if max_order > 0:
+            llh += (
+                (
+                    torch.log(compute_transition_probabilities(self.layers[max_order]))
+                    * self.layers[max_order].data.edge_weight
+                )
+                .sum()
+                .item()
+            )
```
Comment (M-Lampert) [inline src/pathpyG/core/MultiOrderModel.py:452]: Again, some intermediate steps would be nice for readability.

Comment (M-Lampert) [reply src/pathpyG/core/MultiOrderModel.py:452]: Or add the formula as comment as done below

File: src/pathpyG/core/MultiOrderModel.py [hunk 4]
```diff
@@ -278,3 +282,280 @@ def from_DAGs(
...
+                dag_graph.node_sequence.T[0], frequencies.repeat_interleave(dag_graph.ptr[1:] - dag_graph.ptr[:-1])
+            )
+            node_emission_probabilities = counts / counts.sum()
+            llh = torch.mul(torch.log(node_emission_probabilities), counts).sum().item()
+
+        return llh
+
+    def likelihood_ratio_test(
+        self,
+        dag_graph: DataBatch,
```
Comment (M-Lampert) [inline src/pathpyG/core/MultiOrderModel.py:470]:
```suggestion
        dag_graph: Data,
```

File: src/pathpyG/core/MultiOrderModel.py [hunk 5]
```diff
@@ -278,3 +282,280 @@ def from_DAGs(
...
+        x = - 2 * (
+            self.get_mon_log_likelihood(dag_graph, max_order=max_order_null)
+            - self.get_mon_log_likelihood(dag_graph, max_order=max_order)
+        )
+
+        # we calculate the additional degrees of freedom in the alternative model
+        dof_diff = self.get_mon_dof(max_order, assumption=assumption) - self.get_mon_dof(
+            max_order_null, assumption=assumption
+        )
+        # print(x, dof_diff)
```
Comment (M-Lampert) [inline src/pathpyG/core/MultiOrderModel.py:494]:
```suggestion
```

File: src/pathpyG/core/MultiOrderModel.py [hunk 6]
```diff
@@ -278,3 +282,280 @@ def from_DAGs(
...
+
+    def estimate_order(self, dag_data: DAGData, max_order: int = None, significance_threshold: float = 0.01) -> int:
+        """
+        Selects the optimal maximum order of a multi-order network model for the
+        observed paths, based on a likelihood ratio test with p-value threshold of p
+        By default, all orders up to the maximum order of the multi-order model will be tested.
+
+        @param paths: The path statistics for which to perform the order selection
+
+        @param maxOrder: The maximum order up to which the multi-order model shall be tested.
```
Comment (M-Lampert) [inline src/pathpyG/core/MultiOrderModel.py:508]: Wrong doc-string format

Comment (VinsRR) [reply src/pathpyG/core/MultiOrderModel.py]: Fixed in latest commit

File: src/pathpyG/core/MultiOrderModel.py (hunk 3 above)
Comment (VinsRR) [reply src/pathpyG/core/MultiOrderModel.py]: added intermediate steps

File: src/pathpyG/core/MultiOrderModel.py [hunk 7]
```diff
@@ -278,3 +282,280 @@ def from_DAGs(
...
+
+    Args:
+        graph (Graph): pathpy graph object.
+
+    Returns:
+        tensor: Transition probabilities.
+    """
+    weighted_outdegree = compute_weighted_outdegrees(graph)
+    source_ids = graph.data.edge_index[0]
+    return graph.data.edge_weight / weighted_outdegree[source_ids]
```
[the remark this answers is not in the document]
Comment (VinsRR) [reply src/pathpyG/core/MultiOrderModel.py]: They might fit in the `centrality.py` in `algorithms.` But it could also be argued that they should become methods of `Graph` object. Should I put them in `centrality.py` and potentially we'll move them in the future?

File: src/pathpyG/core/MultiOrderModel.py [hunk 8]
```diff
@@ -115,6 +117,7 @@ def aggregate_edge_index(
         if node_sequence.size(1) == 1:
```
[the remark this answers is not in the document]
Comment (VinsRR) [reply src/pathpyG/core/MultiOrderModel.py:117]: moved `unique_nodes, inverse_idx = torch.unique(node_sequence, dim=0, return_inverse=True)` outside the `if ... else` and removed the unecessary lines

File: src/pathpyG/core/MultiOrderModel.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (M-Lampert) [reply src/pathpyG/core/MultiOrderModel.py]: Adding them as methods to `Graph` sounds like a good idea to me. 👍🏼

**Review by IngoScholtes — APPROVED** [review]
Thanks for this implementation! I have now resolved the conflicts with the main branch and migrated the code to the new PathData class. Ready to merge.

File: src/pathpyG/core/MultiOrderModel.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (IngoScholtes) [reply src/pathpyG/core/MultiOrderModel.py]: I have moved those methods to `Graph`

**Comment by IngoScholtes** [comment]
All tests succeeded, I will merge this PR.
